omnigibson/systems/micro_particle_system.py

```diff
@@ -1375,32 +1152,130 @@ def remove_all_particle_instancers(cls):
         """
         cls._sync_particle_instancers(idns=[], particle_groups=[], particle_counts=[])
 
+    @classmethod
+    def create(
+        cls,
+        name,
+        particle_contact_offset,
+        particle_density,
+        **kwargs,
+    ):
+        """
+        Utility function to programmatically generate monolithic fluid system classes.
+
+        Args:
+            name (str): Name of the system
+            particle_contact_offset (float): Contact offset for the generated system
+            particle_density (float): Particle density for the generated system
+            **kwargs (any): keyword-mapped parameters to override / set in the child class, where the keys represent
+                the class attribute to modify and the values represent the functions / value to set
+                (Note: These values should have either @classproperty or @classmethod decorators!)
+
+        Returns:
+            PhysicalParticleSystem: Generated system class
+        """
+
+        # Override the necessary parameters
+        @classproperty
+        def cp_register_system(cls):
+            # We should register this system since it's an "actual" system (not an intermediate class)
+            return True
+
+        @classproperty
+        def cp_particle_contact_offset(cls):
+            return particle_contact_offset
+
+        @classproperty
+        def cp_particle_density(cls):
+            return particle_density
+
+        # Add to any other params specified
+        kwargs["_register_system"] = cp_register_system
+        kwargs["particle_contact_offset"] = cp_particle_contact_offset
+        kwargs["particle_density"] = cp_particle_density
+
+        # Create and return the class
+        return subclass_factory(name=f"{name}System", base_classes=cls, **kwargs)
+
 
 class FluidSystem(PhysicalParticleSystem):
     """
     Particle system class simulating fluids, leveraging isosurface feature in omniverse to render nice PBR fluid
     texture. Individual particles are composed of spheres.
     """
 
+    # Material -- either a MaterialPrim or None if no material is used for this particle system
+    _material = None
+
+    # Color associated with this system (NOTE: external queries should call cls.color)
+    # The default color is blue.
+    _color = np.array([0.0, 0.0, 1.0])
+
+    @classmethod
+    def initialize(cls, simulator):
+        # Run super first
+        super().initialize(simulator=simulator)
+
+        # Create the particle material (only if we're using high-quality rendering since this takes time)
+        # Set custom rendering settings if we're using a fluid isosurface
+        if gm.ENABLE_HQ_RENDERING:
```

Comment:
actually, technically we could have the water material applied but _without_ the isosurface rendering. Should we allow for this case? Might make it easier to interpret what fluid you're seeing, but without the additional overhead of the isosurface rendering

Reply:
What does the water material look like without isosurface rendering?

Reply:
done